File: python/reverse_list.py
class Node:
    def __init__(self, val):
        self.val = val
        self.next = None

# def reverse_list(head):
#   current_node = head 
#   previous_node = None
#   while current_node is not None:
#     next_node = current_node.next
#     current_node.next = previous_node
#     previous_node = current_node
#     current_node = next_node
#   return previous_node

def reverse_list(head, previous_node=None):
  if head is None:
    return previous_node
  next = head.next 
  head.next = previous_node
  return reverse_list(next, head)

a = Node("a")
b = Node("b")
c = Node("c")
d = Node("d")

a.next = b
b.next = c
c.next = d

print(reverse_list(a).val)
print(d.next.val)

File: ruby/reverse_list.rb
class Node
  attr_accessor :val, :next

  def initialize(val)
    @val = val
    @next = nil
  end
end

# def reverse_list(head)
#   current = head 
#   previous = nil 
#   while !current.nil? do 
#     next_node = current.next 
#     current.next = previous 
#     previous = current
#     current = next_node
#   end
#   previous 
# end

def reverse_list(head, previous=nil)
  return previous if head.nil? 
  next_node = head.next
  head.next = previous 
  return reverse_list(next_node, head)
end

a = Node.new("a")
b = Node.new("b")
c = Node.new("c")
d = Node.new("d")

a.next = b
b.next = c
c.next = d

p reverse_list(a).val
p d.next.val